[FEATURE] Rendre le processus d'authentifcation OIDC plus résistant aux retours arrière (PIX-13945)

lego-technix opened the pull request:
## 🌸 Problème

Lorsque l’utilisateur s’authentifie auprès d'un SSO OIDC avec succès, il arrive sur la page d’accueil de Pix. Si l’utilisateur décide de cliquer sur le bouton précèdent de son navigateur alors plusieurs cas de figure problématiques peuvent se produire altérant le parcours utilisateur. Des problèmes peuvent aussi survenir lors d’allers et retours et en faisant des `CTRL + R`.

## 🌳 Proposition

La proposition consiste en plusieurs changements : 
1. La correction d'une race condition lors de la phase de redirection du navigateur de l'utilisateur vers l'authorizationUrl,
2. Lorsque l'utilisateur est déjà authentifié (par Pix), passer outre l'authentification OIDC et rediriger l'utilisateur vers la route `authenticated` (route qui correspond à un utilisateur authentifié), mais sans passer outre la ré-authentification OIDC si l'utilisateur participe à une campagne d'une orga associée à un SSO particulier et que l'utilisateur n’a pas été authentifié par ce SSO,
3. Lorsque l'utilisateur n'est pas authentifié (par Pix), gérer l’erreur `MISSING_OIDC_STATE` en redirigeant l'utilisateur vers la route `authentication.login` (route qui correspond à la page de connexion).

:information_source: Certains SSO refusent qu'on revienne en arrière sur leur page de connexion. Au niveau de Pix on ne peut rien y faire. En fait c'est même la bonne pratique en terme de sécurité. Par contre là où ces SSO pourraient s'améliorer, c'est en ayant alors une page avec une présentation et un texte plus adaptés aux utilisateurs finaux pour mieux expliquer le pourquoi.

## 🐝 Remarques

RAS

## 🤧 Pour tester

### Tests simples sans campagne

1. Se connecter avec les différents SSO, et avec son navigateur faire des retours arrière et des rechargements de page.,
2. Constater que l'utilisateur ne se voit plus présenter de page avec le message `Required cookie "state" is missing` ou `Required "state" is missing in session` et que le parcours utilisateur est logique et fluide.

### Tests complexes avec campagne

1. Créer une orga liée à un SSO1,
2. Créer une campagne pour cette orga,
3. Authentifier un utilisateur auprès d'un SSO2,
4. Avec cet utilisateur aller à l'URL pour accéder à la campagne de l'orga (l'URL de campagne est de la forme `/campagnes/{CODE}`),
5. Constater que l'utilisateur est obligé de s'authentifier auprès du SSO1 (c'est à dire le SSO1 associé à l'orga).

:warning: Pour le cas d'utilisation des orgas associées à un SSO particulier, le comportement n'a pas l'air de bien fonctionner tel qu'il est sur `dev` et tel que déployé en production. Donc si vous trouvez que le fonctionnement que vous rencontrerez dans cette partie est peu satisfaisant, faites les mêmes manipulations sur la recette pour comparer et valider qu'il y a/qu'il n'y a pas de régression.

pix-bot-github commented:
Une fois les applications déployées, elles seront accessibles via les liens suivants :
  * [App (.fr)](https://app-pr11905.review.pix.fr)
  * [App (.org)](https://app-pr11905.review.pix.org)
  * [Orga (.fr)](https://orga-pr11905.review.pix.fr)
  * [Orga (.org)](https://orga-pr11905.review.pix.org)
  * [Certif (.fr)](https://certif-pr11905.review.pix.fr)
  * [Certif (.org)](https://certif-pr11905.review.pix.org)
  * [Junior](https://junior-pr11905.review.pix.fr)
  * [Admin](https://admin-pr11905.review.pix.fr)
  * [API](https://api-pr11905.review.pix.fr/api/)
  * [API MaDDo](https://pix-api-maddo-review-pr11905.osc-fr1.scalingo.io/api/)
  * [Audit Logger](https://pix-audit-logger-review-pr11905.osc-fr1.scalingo.io/api/)

Les variables d'environnement seront accessibles via les liens suivants :
  * [scalingo front](https://dashboard.scalingo.com/apps/osc-fr1/pix-front-review-pr11905/environment)
  * [scalingo api](https://dashboard.scalingo.com/apps/osc-fr1/pix-api-review-pr11905/environment)
  * [scalingo api-maddo](https://dashboard.scalingo.com/apps/osc-fr1/pix-api-maddo-review-pr11905/environment)
  * [scalingo audit-logger](https://dashboard.scalingo.com/apps/osc-fr1/pix-audit-logger-review-pr11905/environment)

lego-technix committed:
refactor(api): more useful error with code and more correct message for 400 BadRequestError during OIDC user authentication

lego-technix committed:
refactor(api): remove duplicate of AlreadyExistingEntityError mapping

lego-technix committed:
test(mon-pix): remove implementation assertions in acceptance test

lego-technix committed:
test(mon-pix): remove useless assert.ok(true) assertions

lego-technix committed:
test(mon-pix): add support for identityProviderCode in SessionStub

P-Jeremy commented:
Test func simple ok ✅

lego-technix committed:
refactor(mon-pix): transform erroneously JSDoc formatted comments

lego-technix committed:
refactor(mon-pix): remove useless toString() on transition parameter

lego-technix committed:
refactor(mon-pix): exit early when identityProvider is not enabled/ready/existing

lego-technix committed:
refactor(mon-pix): rename the 2 OIDC related functions as such

lego-technix committed:
refactor(mon-pix): reorder the 2 OIDC related functions logically

lego-technix committed:
refactor(mon-pix): move code away from useless and misleadingly named _cleanSession

lego-technix committed:
refactor(mon-pix): consistently pass identityProvider argument to functions

instead of identityProviderSlug which is more limited.

lego-technix committed:
fix(mon-pix): fix race condition in redirecting to authorizationUrl

lego-technix committed:
feat(mon-pix): prevent OIDC authentication replay errors when user is authenticated

i.e. 400 BadRequestError 'Required "state" is missing in session'

This is when doing history back and forth and reload.

lego-technix committed:
feat(mon-pix): prevent OIDC authentication replay errors when user is not authenticated

i.e. 400 BadRequestError 'Required "state" is missing in session'

This is when doing history back and forth and reload.

lego-technix committed:
feat(mon-pix): don't loose the /connexion and /connexion/sso-selection URLs from history

which is better for the user to return to when going backwards in the history.

lego-technix committed:
refactor(mon-pix): remove useless return and explicitly use window

lego-technix committed:
feat(mon-pix): add new method replace to location service

lego-technix committed:
test(mon-pix): adapt to new login-oidc route implementation

lego-technix committed:
test(mon-pix): test what the test is about and not the OIDC workflow which is tested in other dedicated acceptance tests

lego-technix committed:
test(mon-pix): fix Mirage values for /oidc/identity-providers

the "id" must have the same value as the "slug" property
cf. api/src/identity-access-management/infrastructure/serializers/jsonapi/oidc-identity-providers.serializer.js

lego-technix committed:
refactor(mon-pix): use the right semantic construct in oidcIdentityProviders service

lego-technix committed:
feat(mon-pix): don't display error details from server to the user

This is because:
* error details originate from server and are untranslated and the error message
  should be translated on the front side.
* error details may contain very technical information not fitted for the user

Libouk commented on mon-pix/tests/helpers/service-stubs.js:
```diff
@@ -32,7 +33,12 @@ export function stubSessionService(owner, sessionData = {}) {
+            user_id: userId,
+            source,
+            access_token: 'access_token!',
+            ...(identityProviderCode && { identityProviderCode }),
```
Pas sûr de lire ça comme il faut, qu'est-ce que ça renvoie ? Possible de l'écrire autrement ?

lego-technix replied:
Ça ajoute une propriété `identityProviderCode` si la variable `identityProviderCode` est `Truthy`. Je suis d'accord que ce n'est pas ce qui est le plus facile à lire. Je vais réécrire de manière à ce que l'intention du code soit plus explicite.

lego-technix replied:
Voilà, le code a été modifié comme ceci, et qui me semble être satisfaisant dans le contexte en question 🙂 : 

```javascript
if (identityProviderCode) {
  Object.assign(this.data.authenticated, { identityProviderCode });
}
```

J'attends ton avis @Libouk 🙂

Libouk commented on mon-pix/app/routes/authentication/login-oidc.js:
```diff
@@ -30,7 +30,7 @@ export default class LoginOidcRoute extends Route {
       this._cleanSession();
 
-      const identityProviderSlug = transition.to.params.identity_provider_slug.toString();
+      const identityProviderSlug = transition.to.params.identity_provider_slug;
```
Il était inutile ? Y'a pas de test concerné ou d'explication à la présence de ce toString ?

lego-technix replied:
Il était complètement archi-inutile. Le `slug` est une `string` de la BDD (table `oidc-providers`) aux consommateurs, toujours manipulé comme une `string`.

Il reste le même `.toString()` inutile dans Pix Admin, mais que j'ai préféré ne pas corriger pour ne pas nous obliger à faire des tests de non-régression sur Pix Admin.

lego-technix committed:
test(mon-pix): factor out unabortedVisit

lego-technix committed:
test(junior): factor out unabortedVisit

pix-bot-github commented:
Une fois les applications déployées, elles seront accessibles via les liens suivants :
  * [App (.fr)](https://app-pr11905.review.pix.fr)
  * [App (.org)](https://app-pr11905.review.pix.org)
  * [Orga (.fr)](https://orga-pr11905.review.pix.fr)
  * [Orga (.org)](https://orga-pr11905.review.pix.org)
  * [Certif (.fr)](https://certif-pr11905.review.pix.fr)
  * [Certif (.org)](https://certif-pr11905.review.pix.org)
  * [Junior](https://junior-pr11905.review.pix.fr)
  * [Admin](https://admin-pr11905.review.pix.fr)
  * [API](https://api-pr11905.review.pix.fr/api/)
  * [API MaDDo](https://pix-api-maddo-review-pr11905.osc-fr1.scalingo.io/api/)
  * [Audit Logger](https://pix-audit-logger-review-pr11905.osc-fr1.scalingo.io/api/)

Les variables d'environnement seront accessibles via les liens suivants :
  * [scalingo front](https://dashboard.scalingo.com/apps/osc-fr1/pix-front-review-pr11905/environment)
  * [scalingo api](https://dashboard.scalingo.com/apps/osc-fr1/pix-api-review-pr11905/environment)
  * [scalingo api-maddo](https://dashboard.scalingo.com/apps/osc-fr1/pix-api-maddo-review-pr11905/environment)
  * [scalingo audit-logger](https://dashboard.scalingo.com/apps/osc-fr1/pix-audit-logger-review-pr11905/environment)
